serialize/src/impls/collections.rs

```diff
@@ -0,0 +1,358 @@
+    ) -> Result<(), SerializationError> {
+        let len = self.len() as u64;
+        len.serialize_with_mode(&mut writer, compress)?;
+        for (k, v) in self {
+            k.serialize_with_mode(&mut writer, compress)?;
+            v.serialize_with_mode(&mut writer, compress)?;
+        }
+        Ok(())
+    }
+
+    fn serialized_size(&self, compress: Compress) -> usize {
+        8 + self
+            .iter()
+            .map(|(k, v)| k.serialized_size(compress) + v.serialized_size(compress))
+            .sum::<usize>()
+    }
+}
+
+impl<K: Valid, V: Valid> Valid for BTreeMap<K, V> {
+    const TRIVIAL_CHECK: bool = K::TRIVIAL_CHECK & V::TRIVIAL_CHECK;
+
+    #[inline]
+    fn check(&self) -> Result<(), SerializationError> {
+        K::batch_check(self.keys())?;
+        V::batch_check(self.values())
+    }
+
+    #[inline]
+    fn batch_check<'a>(batch: impl Iterator<Item = &'a Self>) -> Result<(), SerializationError>
+    where
+        Self: 'a,
+    {
+        let (keys, values): (Vec<_>, Vec<_>) = batch.map(|b| (b.keys(), b.values())).unzip();
+        K::batch_check(keys.into_iter().flatten())?;
+        V::batch_check(values.into_iter().flatten())
+    }
+}
+
+impl<K, V> CanonicalDeserialize for BTreeMap<K, V>
+where
+    K: Ord + CanonicalDeserialize,
+    V: CanonicalDeserialize,
+{
+    /// Deserializes a `BTreeMap` from `len(map) || key 1 || value 1 || ... || key n || value n`.
+    fn deserialize_with_mode<R: Read>(
+        mut reader: R,
+        compress: Compress,
+        validate: Validate,
+    ) -> Result<Self, SerializationError> {
+        let len = u64::deserialize_with_mode(&mut reader, compress, validate)?;
+        (0..len)
+            .map(|_| {
+                Ok((
+                    K::deserialize_with_mode(&mut reader, compress, validate)?,
+                    V::deserialize_with_mode(&mut reader, compress, validate)?,
+                ))
+            })
+            .collect()
+    }
+}
+
+#[cfg(feature = "std")]
+impl<K, V> CanonicalSerialize for std::collections::HashMap<K, V>
+where
+    K: CanonicalSerialize,
+    V: CanonicalSerialize,
+{
+    /// Serializes a `HashMap` as `len(map) || key 1 || value 1 || ... || key n || value n`.
+    fn serialize_with_mode<W: Write>(
+        &self,
+        mut writer: W,
+        compress: Compress,
+    ) -> Result<(), SerializationError> {
+        let len = self.len() as u64;
+        len.serialize_with_mode(&mut writer, compress)?;
+        for (k, v) in self {
+            k.serialize_with_mode(&mut writer, compress)?;
+            v.serialize_with_mode(&mut writer, compress)?;
+        }
+        Ok(())
+    }
+
+    fn serialized_size(&self, compress: Compress) -> usize {
+        8 + self
+            .iter()
+            .map(|(k, v)| k.serialized_size(compress) + v.serialized_size(compress))
+            .sum::<usize>()
+    }
+}
+
+#[cfg(feature = "std")]
+impl<K: Valid, V: Valid> Valid for std::collections::HashMap<K, V> {
+    #[inline]
```

Comment:
no  "const TRIVIAL_CHECK: bool = K::TRIVIAL_CHECK & V::TRIVIAL_CHECK;" for HashMap?

Reply:
Good catch, fixed in latest commit.

Reply:
Good catch, fixed in latest commit